pkg/util/vsphere/vm/cdrom.go

```diff
@@ -0,0 +1,388 @@
+	)
+	deviceChanges = append(deviceChanges, curCdromChanges...)
+
+	// Delete any existing CD-ROM devices that are not in the expected list.
+	// Add them to the device changes with "Remove" operation.
+	curCdroms := util.SelectDevicesByType[*vimtypes.VirtualCdrom](curDevices)
+	for _, cdrom := range curCdroms {
+		if b, ok := cdrom.Backing.(*vimtypes.VirtualCdromIsoBackingInfo); ok {
+			if _, ok := expectedBackingFileNameToCdrom[b.FileName]; ok {
+				continue
+			}
+		}
+
+		deviceChanges = append(deviceChanges, &vimtypes.VirtualDeviceConfigSpec{
+			Device:    cdrom,
+			Operation: vimtypes.VirtualDeviceConfigSpecOperationRemove,
+		})
+	}
+
+	return deviceChanges, nil
+}
+
+// UpdateConfigSpecCdromDeviceConnection updates the connection state of the
+// VM's existing CD-ROM devices to match what specifies in VM.Spec.Cdrom list.
+func UpdateConfigSpecCdromDeviceConnection(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	config *vimtypes.VirtualMachineConfigInfo,
+	configSpec *vimtypes.VirtualMachineConfigSpec) error {
+	var (
+		cdromSpec                    = vmCtx.VM.Spec.Cdrom
+		curDevices                   = object.VirtualDeviceList(config.Hardware.Device)
+		backingFileNameToCdromSpec   = make(map[string]vmopv1.VirtualMachineCdromSpec, len(cdromSpec))
+		backingFileNameToCdromDevice = make(map[string]vimtypes.BaseVirtualDevice, len(cdromSpec))
+	)
+
+	for _, specCdrom := range cdromSpec {
+		imageRef := specCdrom.Image
+		if imageRef == nil {
+			return fmt.Errorf("image ref is nil for CD-ROM: %q", specCdrom.Name)
+		}
+
+		cdrom, bFileName, err := getCdromAndBackingByImgRef(vmCtx, client, *imageRef, curDevices)
+		if err != nil {
+			vmCtx.Logger.Error(err, "failed to get CD-ROM device by image ref", "imageRef", *imageRef)
+			return err
+		}
+
+		if cdrom == nil {
+			// This could happen if the VM spec has a new CD-ROM device, or the
+			// existing CD-ROM device's backing has been changed. The former
+			// situation should be denied by the VM validating webhook.
+			return fmt.Errorf("no CD-ROM is found for image ref %s", *imageRef)
+		}
+
+		backingFileNameToCdromSpec[bFileName] = specCdrom
+		backingFileNameToCdromDevice[bFileName] = cdrom
+	}
+
+	curCdromChanges := updateCurCdromsConnectionState(
+		backingFileNameToCdromSpec,
+		backingFileNameToCdromDevice,
+	)
+	configSpec.DeviceChange = append(configSpec.DeviceChange, curCdromChanges...)
+
+	return nil
+}
+
+// getCdromAndBackingByImgRef returns the CD-ROM device and its backing file
+// name from the given image reference.
+func getCdromAndBackingByImgRef(
+	vmCtx pkgctx.VirtualMachineContext,
+	client ctrlclient.Client,
+	imageRef vmopv1.VirtualMachineImageRef,
+	curDevices object.VirtualDeviceList) (vimtypes.BaseVirtualDevice, string, error) {
+
+	fileName, err := getIsoFilenameFromImageRef(vmCtx, client, imageRef)
+	if err != nil {
+		vmCtx.Logger.Error(err, "failed to get ISO file name from image ref", "imageRef", imageRef)
```

Comment:
Please see my previous comments about returning a contextual error instead of logging the error and returning it.